prohibit the use of T.{untyped,noreturn} as values

froydnj opened the pull request:
This change ensures that `T.{untyped,noreturn}` are only used as types, not values.

### Motivation
<!-- Why make this change? Describe the problem, not the solution. This can also be a link to an issue. -->

Fixes #3426 
Fixes #3428 

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

jez commented on ast/Helpers.h:
```diff
@@ -355,10 +355,14 @@ class MK {
+    static TreePtr UntypedValue(core::LocOffsets loc) {
+        return Unsafe(loc, Nil(loc));
+    }
+
```
My preference would have been to keep `MK::Untyped` named as is and to call this one `UnsafeNil`, which makes it more clear at the call site what's going on.

`T.unsafe(nil)` in DSL passes is generally a smell and we've been trying to move away from them, towards either rewriting it to what the runtime would actually do, or to a raise:

https://github.com/sorbet/sorbet/pull/2883/files#diff-0ebfafdc5884d2abe7cce00031996991cfb2bcdc67e666a7a34dab8052c4f2d2L113-R113

(Alternatively, if you wanted to try changing the places using your `MK::UntypedValue` helper to use `RaiseUnimplemented` that would be another way forwards.)

jez commented on test/testdata/infer/all_bug.rb:
```diff
@@ -18,6 +18,7 @@ def self.test(x)
 end
 
-T.assert_type!(Test.test(T.untyped), M)
+T.assert_type!(Test.test(T.untyped), M) # error: This code is unreachable
+                       # ^^^^^^^^^ error: Expected `T.all(M, T)` but found `<Type: T.untyped>` for argument `x`
```
My guess is that the author's original intention with this bug was to use `T.unsafe(nil)` here.

jez replied:
i.e., we should hope that we didn't change the behavior of the `This code is unreachable error` because we didn't change `T.all`

jez commented on test/testdata/generic_noreturn.rb:
```diff
@@ -0,0 +1,8 @@
+end
+
+x = MyGeneric[T.noreturn].new
+T.reveal_type(x) # error: Revealed type: `MyGeneric[T.noreturn]`
```
🎉

jez reviewed (approved):
One suggestion on naming (because half the point of this change is to make it so that `Untyped` always implies "we're talking about types" and never "we're talking about values")

froydnj commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_IHvMwfNZNU8IUC

froydnj commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_IHvfAIF4fTTtpw

froydnj commented:
> One suggestion on naming (because half the point of this change is to make it so that `Untyped` always implies "we're talking about types" and never "we're talking about values")

I made the `UntypedNil` change, reverted `UntypedType`, and changed the `all_bug.rb` test to use `T.unsafe(nil)`.  I might have a go at fixing things to use `RaiseUnimplemented` in a different PR tomorrow.

froydnj commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_IHw7Zsab5USzLS
